### ISSUE TITLE
out of memory error on a huge machine

### ISSUE THREAD

**Issue by guyisra** [osmcode/osmium-tool#135]
trying to sort an osm file of about 30G, on a huge machine in AWS with 900+ GB RAM
it reaches 100% and still exists on out of memory error
The machine has 30x more memory than the file, yet still running out

using osmium 1.8.0 

Is there a way to debug this further?

**Comment by guyisra** [comment]
Thanks!

That's what I am trying but it takes a long time to sort the nodes file.

Are there any benchmarks for the time it should take to sort?

**Comment by joto** [comment]
I have no idea how long this "should" take. I haven't tried it with anything of the size you are talking about here. You can try with smaller files and interpolate. Not sure how good those numbers will be though.

I have added an option to `osmium sort` that does the sort for nodes, ways, and relations separately (`--strategy=multipass`). Maybe this will help you. But from my tests, the memory savings aren't that large, because the memory needed for the nodes overwhelmes everything else.

I have also added some output about used buffers, it should show up when you are using `-v`. This could help us debugging. Similar output is available from `osmium fileinfo -e` now. Please report the numbers you are seeing with those commands.

From my tests I am seeing that the use of the buffers is particularly inefficient for nodes. There is a huge overhead due to the way buffers are re-allocated when they run out of space. I am looking into this and can maybe improve this.

**Comment by guyisra** [comment]
unfortunately, using the multipass strategy errors out on OOM at 31% (on a machine with ~940GB Mem)

Using fileinfo on the entire osm.pbf file gives these buffers:
  Number of buffers: 1072983 (avg 3999 objects per buffer)
  Sum of buffer sizes: 369085956520 (369.085 GB)
  Sum of buffer capacities: 2250296524800 (2250.3 GB, 16% full)

Using fileinfo on the nodes file (after using cat) gives these buffers:
  Number of buffers: 464051 (avg 7999 objects per buffer)
  Sum of buffer sizes: 230206656144 (230.206 GB)
  Sum of buffer capacities: 973185482752 (973.185 GB, 24% full)

Does this mean I need 973 GB at least?

**Comment by joto** [comment]
Yes, it means you need 973 GB plus some more for the index, another `8 * larges_node_id`. The problem is that those buffers are only 24% full, I have an idea how to improve this, but at the moment there is no easy fix.

**Comment by joto** [comment]
Here is a workaround you can try: There is an `osmium merge` command which can merge several sorted files into one. If you can split up your data somehow, you can sort each part and then use `osmium merge` to merge the sorted parts. There is no way to split an OSM file with Osmium directly, but you could use `osmium cat -f opl` to write out the OSM files in OPL format and then use the Unix `split` command to split this into multiple files. Of course the resulting OPL file is going to be huge...

**Comment by guyisra** [comment]
that's the method I'm currently using. But using cat with -f osm.pbf

**Comment by joto** [comment]
Not sure I understand. You can't use the Unix `split` on a PBF file, but you can use it on an OPL file.

**Comment by guyisra** [comment]
this is sort of what I do

```
#! /bin/bash 

extract(){
  type=$1
  file=$2
  out=$3
  osmium cat -f osm.pbf -t "$type" "$file" -o "$out" -O &
}

sortType(){
  input=$1
  output=$2
  osmium sort "$input" -o "$output" -O &
}

sortOsm(){
  inputFileName=$1
  nodesFileName=nodes.osm.pbf
  waysFileName=ways.osm.pbf
  relationsFileName=relations.osm.pbf

  extract node "$inputFileName" "$nodesFileName"
  extract way "$inputFileName" "$waysFileName"
  extract relation "$inputFileName" "$relationsFileName"

  wait 

  sortedNodes=nodes-sorted.osm.pbf
  sortedWays=ways-sorted.osm.pbf
  sortedRelations=relations-sorted.osm.pbf

  sortType "$nodesFileName" "$sortedNodes"
  sortType "$waysFileName" "$sortedWays" 
  sortType "$relationsFileName" "$sortedRelations" 

  wait

  osmium merge "$sortedNodes" "$sortedWays" "$sortedRelations" -o "$2" -O
}
```

**Comment by joto** [comment]
Okay, we are misunderstanding each other. First, running the `sortType`s in parallel doesn't make much sense when you don't have enough memory. Second, you don't need `merge` in the last step, because `cat` will do fine, because sorted OSM files have nodes first, then ways, then relations. But all of this is beside the point, because you will run out of memory regardless.

But you can solve this by splitting up your file even more, ie. not having all nodes in one file, but in several files. Either whichever process is creating this data can write out the data in pieces or you could use the OPL format (https://osmcode.org/opl-file-format/) which is a lines-of-text-based format that you can `split` (or even `sort`) with unix tools.

**Comment by guyisra** [comment]
ah I see, thanks!
after I have (for example) nodes-1.opl and nodes-2.opl and I sort each one, how will osmium know that the merge of these will be sorted?
nodes-1.opl can have
```
n1 ...
n3 ...
```
and nodes-2.opl can have
```
n2 ...
n4 ...
```

so by themselves they are sorted, but when merged the nodes won't be in order

**Comment by joto** [comment]
Please read the manpage for `osmium merge`.

**Comment by joto** [comment]
I have experimented with some changes that reduce memory consumption dramatically, often to less that half. The price is about 1-3% increase in runtime. This is preliminary work and not well tested, but if you want to test it, it is in the [`exp-nested-buffers` branch](https://github.com/osmcode/libosmium/tree/exp-nested-buffers).

**Comment by guyisra** [comment]
thanks! I'll test this soon

**Comment by guyisra** [comment]
it took 104 minutes with  Peak memory used: 560488 MBytes (on a ~38GB file)
So it as you wrote.

It worked pretty smoothly.

I've noticed a 2 bytes difference between the file sorted by split and merge and the exp-nested-buffers sort. It's not a big deal, just FYI

**Comment by joto** [comment]
I have now committed a modified version of this code to libosmium master. The performance decrease is now even smaller and the memory overhead also smaller. Closing here.
